### PR TITLE
The dynamic backend feature causes crashes

### PR DESCRIPTION
**Pull request by dmatetelki**
- Added "R" to the short options
- Added "backend-refresh" to config file parsing
- Made "backend-refresh" default to 0.
- Corrected memcpy size parameters.
- Using freeaddrinfo on updates
- Better comparison of 2 addrinfos

### CODE REVIEW

**Comment by klenan** [comment]
You beat me to it, thanks! Quick testing showed no regression so far, will test more thoroughly under high load and report here later.

One thing I'm still missing: Updates to the manual page. But as other bits such as `-M` and seperate key files are also lacking documentation/mentioning as of now I'm unsure as to whether this should go into a seperate PR. Thoughts?

File: src/hitch.c [hunk 1]
```diff
@@ -2474,7 +2475,7 @@ handle_mgt_rd(struct ev_loop *loop, ev_io *w, int revents)
 		backaddr->ai_family = wu.payload.addr.family;
 		backaddr->ai_addrlen = wu.payload.addr.len;
 		memcpy(backaddr->ai_addr, &wu.payload.addr.addr,
-		       sizeof(struct sockaddr_storage));
```
Comment (daghf) [inline src/hitch.c]: Wouldn't this overflow if we had an IPv4 address that is refresh to an IPv6 address?

File: src/hitch.c [hunk 2]
```diff
@@ -2766,27 +2767,47 @@ drop_privileges(void)
 #endif
 }
 
+/* Returns 0 on match, 1 if differ */
+int
```
Comment (daghf) [inline src/hitch.c]: make this static

File: src/hitch.c [hunk 3]
```diff
@@ -2766,27 +2767,47 @@ drop_privileges(void)
...
+addrinfocmp(struct addrinfo *a, struct addrinfo *b)
```
Comment (daghf) [inline src/hitch.c]: constify the parameters

File: src/hitch.c [hunk 4]
```diff
@@ -2766,27 +2767,47 @@ drop_privileges(void)
...
+	    a->ai_socktype != b->ai_socktype ||
+	    a->ai_protocol != b->ai_protocol ||
+	    a->ai_addrlen  != b->ai_addrlen)
+		return 1;
+
+	/* disregard ai_canonname and ai_next */
+	return memcmp(a->ai_addr, b->ai_addr, a->ai_addrlen) == 0 ? 0 : 1;
+}
+
 int
```
Comment (daghf) [inline src/hitch.c]: Should be static

File: src/hitch.c [hunk 5]
```diff
@@ -2766,27 +2767,47 @@ drop_privileges(void)
...
 	if (gai_err != 0) {
 		ERR("{getaddrinfo-backend}: %s\n", gai_strerror(gai_err));
 		exit(1);
 	}
-	if (backaddr == 0 ||
-	    memcmp(addrinfo->ai_addr, backaddr->ai_addr,
-		   sizeof(struct sockaddr_storage)) != 0) {
-		backaddr = addrinfo;
+	if (backaddr == 0 || addrinfocmp(result, backaddr) != 0) {
+		if (backaddr == 0)
```
Comment (daghf) [inline src/hitch.c]: Was this supposed to be `!= NULL` instead?

Comment (klenan) [reply src/hitch.c]: Yes, `if (backaddr != NULL || ...) {`.

File: src/hitch.c (hunk 1 above)
Comment (klenan) [reply src/hitch.c]: `struct sockaddr_storage` is designed to be family agnostic, see [RFC 2553 Section 3.10](https://tools.ietf.org/html/rfc2553#section-3.10).

**Comment by klenan** [comment]
https://github.com/varnish/hitch/blob/e5e23cb53691a023aefec4ca2d21633d2019cab9/src/hitch.c#L3503 still fails with `BADF`:

```
$ git describe
hitch-1.4.6-21-ge5e23cb
$ fgrep hitch /etc/hosts
$ ./hitch -q -uhitch -f[::1]:443 -b[hitch-tls.org]:80 -R1 ./cert.pem &
[1] 30314
$ 20171101T202621.994697 [30314] {core} hitch 1.4.6 starting
20171101T202621.999570 [30314] {core} Loading certificate pem files (1)
20171101T202622.024324 [30315] {core} Process 0 online
20171101T202622.024592 [30314] {core} hitch 1.4.6 initialization complete

$ echo ::1 hitch-tls.org >>/etc/hosts; sleep 2
$ sed -i /hitch/s,::1,127.0.0.1, /etc/hosts; echo

$ 20171101T202713.040000 [30314] WARNING: {core} Unable to notify worker 30315 with changed backend address (Bad file descriptor).
20171101T202713.041025 [30314] {core} Child 30315 exited with status 0.
20171101T202713.041441 [30329] {core} Process 0 online
```

File: src/hitch.c (hunk 1 above)
Comment (daghf) [reply src/hitch.c]: Hi @klenan, the problem here is that we need to consider the storage of what is being copied into. `backaddr->ai_addr` is either a `struct sockaddr_in` or a `struct sockaddr_in6`, both significantly smaller than the size of a `struct sockaddr_storage`.

The attempted fix doesn't work either, which is what my previous comment refers to.

The proper fix is probably to change the backaddr definition, possibly by making it a `struct sockaddr_storage`, capable of holding both IPv6 and IPv4 sockaddrs.

Comment (Dridi) [reply src/hitch.c]: That, or adopting `struct suckaddr`...

**Comment by dmatetelki** [comment]
@Dridi : I thought you were joking with `suckaddr`, but life is stranger than fiction :P

**Comment by dmatetelki** [comment]
oh, it crashes at:

hitch.c:2484
`backaddr = VSA_Clone(wu.payload.addr);`

to be continued...

**Comment by daghf** [comment]
`VSA_Clone` takes a `suckaddr` as input, I think `VSA_Malloc` might be what you're looking for

File: src/foreign/vrt.h [hunk 6]
```diff
@@ -0,0 +1,377 @@
+/*-
```
Comment (daghf) [inline src/foreign/vrt.h]: We definitely don't want to include vrt.h - could we slim down the VSA bits ever so slightly to not depend on this?

File: src/hitch.c [hunk 7]
```diff
@@ -1456,7 +1451,11 @@ create_frontend(const struct front_arg *fa)
 static int
 create_back_socket()
 {
-	int s = socket(backaddr->ai_family, SOCK_STREAM, IPPROTO_TCP);
+	socklen_t len;
+	const void *addr = VSA_Get_Sockaddr(backaddr, &len);
+
```
Comment (daghf) [inline src/hitch.c]: Add `AN(addr);`

File: src/hitch.c [hunk 8]
```diff
@@ -1659,7 +1661,11 @@ handle_connect(struct ev_loop *loop, ev_io *w, int revents)
 	(void)revents;
 	CAST_OBJ_NOTNULL(ps, w->data, PROXYSTATE_MAGIC);
 
-	t = connect(ps->fd_down, backaddr->ai_addr, backaddr->ai_addrlen);
+	socklen_t len;
+	const void *addr = VSA_Get_Sockaddr(backaddr, &len);
+
```
Comment (daghf) [inline src/hitch.c]: `AN(addr);`

File: src/hitch.c [hunk 9]
```diff
@@ -2471,10 +2498,12 @@ handle_mgt_rd(struct ev_loop *loop, ev_io *w, int revents)
...
-		backaddr->ai_family = wu.payload.addr.family;
-		backaddr->ai_addrlen = wu.payload.addr.len;
-		memcpy(backaddr->ai_addr, &wu.payload.addr.addr,
-		       sizeof(struct sockaddr_storage));
+		if (backaddr != NULL)
+			free(backaddr);
+
+		socklen_t len;
+		const void *addr = Get_Sockaddr(&(wu.payload.addr), &len);
+		backaddr = VSA_Malloc(addr, len);
```
Comment (daghf) [inline src/hitch.c]: Add a `AN(VSA_Sane(backaddr));` here?

File: src/hitch.c [hunk 10]
```diff
@@ -2471,10 +2498,12 @@ handle_mgt_rd(struct ev_loop *loop, ev_io *w, int revents)
 		(worker_state == WORKER_EXITING) ? "EXITING" : "ACTIVE");
 	}
 	else if (wu.type == BACKEND_REFRESH) {
-		backaddr->ai_family = wu.payload.addr.family;
-		backaddr->ai_addrlen = wu.payload.addr.len;
-		memcpy(backaddr->ai_addr, &wu.payload.addr.addr,
-		       sizeof(struct sockaddr_storage));
+		if (backaddr != NULL)
```
Comment (daghf) [inline src/hitch.c]: `free(NULL)` is totally legitimate in C, so this `if (backaddr != NULL)` test could just be killed

File: src/hitch.c [hunk 11]
```diff
@@ -2768,25 +2797,36 @@ drop_privileges(void)
 
 int
```
Comment (daghf) [inline src/hitch.c]: static?

Comment (dmatetelki) [reply src/hitch.c]: you'd like `void drop_privileges(void)` to be static?

Comment (daghf) [reply src/hitch.c]: I'm referring to `get_backaddr_info()`.

That said, `drop_privileges` should probably be too. :-)

File: hitch.man.rst [hunk 12]
```diff
@@ -50,6 +50,7 @@ Command line arguments
   -n  --workers=NUM          Number of worker processes (Default: 1)
   -B  --backlog=NUM          Set listen backlog size (Default: 100)
   -k  --keepalive=SECS       TCP keepalive on client socket (Default: 3600)
+  -R  --backendrefresh=SECS  Periodic backend IP lookup, 0 to disable (Default: 0)
```
Comment (klenan) [inline hitch.man.rst]: I suggest `--backend-refresh` to be consistent with the `backend-refresh` configuration parameter just like `prefer-server-ciphers`, `write-proxy*` et al. do.

File: src/configuration.c [hunk 13]
```diff
@@ -1098,7 +1098,7 @@ config_print_usage_fd(char *prog, FILE *out)
 	fprintf(out, "  -n  --workers=NUM          Number of worker processes (Default: %ld)\n", cfg->NCORES);
 	fprintf(out, "  -B  --backlog=NUM          Set listen backlog size (Default: %d)\n", cfg->BACKLOG);
 	fprintf(out, "  -k  --keepalive=SECS       TCP keepalive on client socket (Default: %d)\n", cfg->TCP_KEEPALIVE_TIME);
-	fprintf(out, "  -R  --backendrefresh=SECS  Periodic backend IP lookup (Default: %d, 0 to disable)\n", cfg->BACKEND_REFRESH_TIME);
+	fprintf(out, "  -R  --backendrefresh=SECS  Periodic backend IP lookup, 0 to disable (Default: %d)\n", cfg->BACKEND_REFRESH_TIME);
```
Comment (klenan) [inline src/configuration.c]: Same here, of course.

**Comment by dmatetelki** [comment]
Abandoning for #230

**Pull request closed without merging**